rfcs/2021-06-01-multilinks.md

```diff
@@ -0,0 +1,354 @@
+
+# Guide-level explanation
+
+Links are the most important feature in Subtext. By allowing you to reference other documents, you can compose hypertext documents from many smaller documents.
+
+Subtext link blocks allow you to reference other files by path, or by URL. Files can be of any type, even other Subtext documents! Subtext clients typically render a link block as a transclude, displaying all or part of the content of the link in-place. What this looks like will depend on the file type and the Subtext client. For example:
+
+- A link to an image may be displayed in-place
+- A link to a video may be displayed with playback controls
+- A link to CSV file may be rendered as a table
+- A link to a Subtext file may render all or some of the Subtext file within another Subtext file.
+
+Think of Subtext as hypertext glue that lets you compose larger documents from smaller documents.
+
+A basic link is just a link sigil (`&`) followed by a path or URL. These are all valid link blocks:
+
+```
+& example.png
+& ../example.png
+& http://example.com
+```
+
+Links may also include a label. Labels are a people-friendly discription of the link. They can be used as a text-based fallback when a client doesn't know how to transclude a given type of file. Labels come after the URL, like this:
+
+```
+http://example.com Example label
+```
+
+You can also write a link with only a label. We call this a wikilink. Clients may use the label to decide on a resource they think matches best. This is analogous to the way `[[wiklink]]` tags work in [Wiki markup](https://en.wikipedia.org/wiki/Help:Wikitext).
+
+```
+& Example label
+```
+
+Links can also contain multiple URLs. This allows you to describe multiple places where a resource may be found. This can be useful when sharing content from newer protocols that may not be widely supported yet, such as `ifps` or `dat`. In these cases you can offer multiple URLs and the client can choose whichever one it understands.
+
+Multiple urls are separated by a pipe character (`|`).
+
+```
+& http://example.com|ipfs://example|dat://example
+```
+
+You can also have multiple URLs and a label:
+
+```
+& http://example.com|ipfs://example|dat://example Example label
+```
+
+# Reference-level explanation
+
+Link blocks start with `& `. They reference other files within the flow of a Subtext document. Any kind of file can be linked, including other Subtext documents.
+
+Links are the most important feature in Subtext. By allowing you to reference other documents, you can compose hypertext documents from many smaller documents.
+
+Links can include zero or more pipe-delimited URLs, and an optional text lable. They often take one of the following forms:
+
+- URL
+- URL with label
+- Multiple URLs
+- Multiple URLs with label
+- Label only (wikilink)
+
+The following are all valid URLs
+
+```
+& example.png
```

Comment:
Is this a URL or a wikilink?

Reply:
oof will have to think about how to disambiguate

Reply:
Ideas I've come up with so far:

- Require paths to start with a path component (`/`, `./`, `../`, `scheme://`)
  - Cons: unambiguous but requires a bit of logic to work out what you are looking at.
- Have separate wikilink block sigil, such as `?` or `+`.
  - Cons: feels redundant, and I like the idea of Wikilinks just being a special unqualified kind of link.

Others?

Reply:
IA Writer offers a bit of prior art to consider: https://github.com/gordonbrander/subtext/blob/multilink/rfcs/2021-06-01-multilinks.md#ia-transclusion-syntax. Its transclude syntax is:

```
./example.txt "Caption"
```

- Path must begin on new line, start with `./`, `../`, or `/` and include a file extension (for example `.txt`) to be recognized as a valid path.
- Caption must be wrapped in quotes or parens.

Reply:
I've implemented idea 1 in Python as a way of testing it out https://github.com/gordonbrander/subtext/blob/multilink/parsers/subtext.py. It's reasonably simple. Leaning in this direction.

Reply:
I'm going with 1.